torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -1051,6 +1052,91 @@ int listClear(Stack& stack) {
   return 0;
 }
 
+template <typename TList, typename TElement>
+int listInsert(Stack& stack) {
+  TList list;
+  int64_t idx;
+  TElement elem;
+  pop(stack, list, idx, elem);
+
+  auto& elements = list->elements();
+  const int64_t list_size = elements.size();
+  const int64_t normalized_idx = normalizeIndex(idx, list_size);
+
+  if (normalized_idx < 0 || normalized_idx >= list_size) {
+    if (idx < 0) {
+      elements.insert(elements.begin(), elem);
+    } else {
+      elements.push_back(elem);
+    }
+  } else {
+    elements.insert(elements.begin() + normalized_idx, elem);
+  }
+
+  return 0;
+}
+
+template <typename TList, typename TElement>
+int listRemove(Stack& stack) {
+  TList list;
+  TElement elem;
+  pop(stack, list, elem);
+
+  auto& elements = list->elements();
+  auto pos = std::find(elements.begin(), elements.end(), elem);
+
+  if (pos != elements.end()) {
+    elements.erase(pos);
+  } else {
+    AT_ERROR("list.remove(x): x not in list");
+  }
+
+  return 0;
+}
+
+template <>
+int listRemove<Shared<TensorList>, at::Tensor>(Stack& stack) {
+  Shared<TensorList> list;
+  at::Tensor elem;
+  pop(stack, list, elem);
+
+  auto& elements = list->elements();
+  auto pos = std::find_if(elements.begin(), elements.end(), [elem](const at::Tensor& b) {
+    const auto cmp_result = elem.eq(b);
+    if (!cmp_result.is_nonzero()) {
+      return false;
+    }
+    return true;
+  });
+
+  if (pos != elements.end()) {
+    elements.erase(pos);
+  } else {
+    AT_ERROR("list.remove(x): x not in list");
+  }
+
+  return 0;
+}
+
+template <>
+int listRemove<Shared<GenericList>, IValue>(Stack& stack) {
+  Shared<GenericList> list;
+  IValue elem;
+  pop(stack, list, elem);
+
+  auto& elements = list->elements();
+  auto pos = std::find_if(elements.begin(), elements.end(), [elem](const IValue& b) {
+    return b.isAliasOf(elem);
```

Comment:
@eellison is this the correct way to check for equality of IValues?

Reply:
Hi, 

Python list remove checks each element of the list for value equality to the input. IsAliasOf checks reference equality. Basically the difference between using `a == b` and `a is b`. You probably are going to want to do some refactoring and use our existing implementations of the equality operator "aten::eq." Thanks for the PR, looks good!!

Reply:
In which case would the `IValue` case be called?

Reply:
IsAliasOf is in the GenericList implementation, which means all the lists we don't have existing specializations for - everything but tensors, ints, floats, and bools. For example, a list of int lists. Pretty sure [[.1, .2], [.5], [1.0]].remove([.5]) would remove an element in python but would not work currently in torchscript. 

FYI: a [writeup](https://github.com/pytorch/pytorch/blob/f42f7c95b301ba02df04f722393e5b5da4d41718/torch/csrc/jit/README.md#ivalue) on ivalues & other technicals

Reply:
Thanks for the help, didn't think of that case 👍 The document looks really helpful btw!
I couldn't really come up with a solution on how to use aten::eq, so my idea currently is to move the remove op out of CREATE_MUTABLE_LIST_OPS and move it to its own CREATE_REMOVE_OP, which is defined for TensorList and GenericList.
And then do something along the lines of [ListConstruct](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/register_prim_ops.cpp#L786-L793).
Maybe I'm overcomplicating things but I can't think of a better solution atm.

So any more help would be much appreciated.

Reply:
Yea agreed - If you were implementing GenericList you would probably doing something like ListConstruct where you dispatch to the right function based on the input type of the list.

But aten::eq isn't implemented currently for GenericLists, it's only implemented for our specialized lists. I think it's probably better if you remove GenericList from this PR and leave it for a follow up.